src/validators/arguments_v3.rs

```diff
@@ -0,0 +1,654 @@
+                        }
+                        Err(ValError::LineErrors(line_errors)) => {
+                            errors.extend(
+                                line_errors
+                                    .into_iter()
+                                    .map(|err| lookup_path.apply_error_loc(err, self.loc_by_alias, &parameter.name)),
+                            );
+                        }
+                        Err(err) => return Err(err),
+                    }
+                }
+                (None, None) => {
+                    if let Some(value) = parameter
+                        .validator
+                        .default_value(py, Some(parameter.name.as_str()), state)?
+                    {
+                        if parameter.mode == ParameterMode::PositionalOnly {
+                            output_args.push(value);
+                        } else {
+                            output_kwargs.set_item(PyString::new(py, parameter.name.as_str()).unbind(), value)?;
+                        }
+                    } else {
+                        // Required and no default, error:
+                        match parameter.mode {
+                            ParameterMode::PositionalOnly => {
+                                errors.push(ValLineError::new_with_loc(
+                                    ErrorTypeDefaults::MissingPositionalOnlyArgument,
+                                    original_input,
+                                    index,
+                                ));
+                            }
+                            ParameterMode::PositionalOrKeyword => {
+                                errors.push(lookup_key.error(
+                                    ErrorTypeDefaults::MissingArgument,
+                                    original_input,
+                                    self.loc_by_alias,
+                                    &parameter.name,
+                                ));
+                            }
+                            ParameterMode::KeywordOnly => {
+                                errors.push(lookup_key.error(
+                                    ErrorTypeDefaults::MissingKeywordOnlyArgument,
+                                    original_input,
+                                    self.loc_by_alias,
+                                    &parameter.name,
+                                ));
+                            }
+                            _ => unreachable!(),
+                        }
+                    }
+                }
+            }
+        }
+
+        // if there are args check any where index > positional_params_count since they won't have been checked yet
+        if let Some(args) = args_kwargs.args() {
+            let len = args.len();
+            if len > self.positional_params_count {
+                if let Some(var_args_param) = self.parameters.iter().find(|p| p.mode == ParameterMode::VarArgs) {
```

Comment:
If we validated the structure of the parameters at build time, we might be able to just know this statically without having to iterate to find it.

Reply:
So you mean having an extra field on the validator struct, as an `Option<Parameter>`?

Reply:
I was thinking so, yes. Might even want something like

```rust
positional_args: Vec<Parameter>,
var_args: Option<Parameter>,
keyword_only_args: Vec<Parameter>,
var_kwargs: Option<Parameter>,
```

Reply:
.. but also fine to leave as follow up.